src/ProgressLogging.jl

```diff
@@ -3,13 +3,168 @@ module ProgressLogging
 # Use README as the docstring of the module:
 @doc read(joinpath(dirname(@__DIR__), "README.md"), String) ProgressLogging
 
-export @progress, @withprogress, @logprogress
+export @progress, @progressid, @withprogress, @logprogress
 
 using Base.Meta: isexpr
+using UUIDs: UUID, uuid4, uuid5
 using Logging: Logging, @logmsg, LogLevel
 
 const ProgressLevel = LogLevel(-1)
 
+"""
+    ProgressLogging.ROOTID
+
+This is used as `parentid` of root [`Progress`](@ref)es.
+"""
+const ROOTID = UUID(0)
+
+"""
+    ProgressLogging.Progress(id, [fraction]; [parentid, name, done])
+
+# Usage: Progress log record provider
+
+Progress log record can be created by using the following pattern
+
+```julia
+id = uuid4()
+try
+    @info Progress(id)  # create a progress bar
+    # some time consuming task
+    # ...
+    @info Progress(id, 0.1)  # update progress to 10%
+    # ...
+finally
+    @info Progress(id, done = true)  # close the progress bar
+end
+```
+
+It is recommended to use [`@withprogress`](@ref),
+[`@logprogress`](@ref), and optionally [`@progressid`](@ref) to create
+log records.
+
+# Usage: Progress log record consumer (aka progress monitor)
+
+It is recommended to use [`ProgressLogging.asprogress`](@ref) instead
+of checking `message isa Progress`.  Progress monitors can retrieve
```

Comment:
Should we recommending `asprogress` over `message isa Progress`?  With this approach, progress monitors can/have to use duck-typing to handle progress event.  A nice effect is that it exposes a very small set of API for progress monitors.  In fact, I need to import only `asprogress` in TerminalLoggers.jl:

import: https://github.com/c42f/TerminalLoggers.jl/pull/13/commits/e9676d825889147a9629cb48559caaefd64137ec#diff-ea4f99d2ca25dd86c4f33f869af6f1b7R11
usage: https://github.com/c42f/TerminalLoggers.jl/commit/e9676d825889147a9629cb48559caaefd64137ec#diff-970cebd9168130c35d89887db7899b60R197-R199

Of course, a disadvantage (from some point of view) is that it exposes properties as public API.

Reply:
That's fine imho. If someone *really* wanted to avoid the import they could also check for the existence of `Progress`'s fields directly.